File: frontend/src/utils/validators.js
// validators.js placeholder
// src/utils/validators.js

import {
  STUDENT_LEVELS,
  HOSTELS,
  UG_YEARS,
  EMAIL_REGEX,
  MAX_IMAGE_SIZE_MB,
} from "./constants";

// ---------------- STUDENT LOGIN VALIDATION ---------------- //

export function validateStudentLogin(formData) {
  const errors = [];

  const { name, level, branch, year, department, hostel } = formData || {};

  if (!name || !name.trim()) {
    errors.push("Name is required.");
  }

  if (!STUDENT_LEVELS.includes(level)) {
    errors.push("Level must be UG or PG.");
  }

  if (!HOSTELS.includes(hostel)) {
    errors.push("Please select a valid hostel.");
  }

  if (level === "UG") {
    if (!branch || !branch.trim()) {
      errors.push("Branch is required for UG students.");
    }
    const yrNum = Number(year);
    if (!UG_YEARS.includes(yrNum)) {
      errors.push("Year must be between 1 and 4 for UG.");
    }
  }

  if (level === "PG") {
    if (!department || !department.trim()) {
      errors.push("Department is required for PG students.");
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
}

// ---------------- DOCTOR LOGIN VALIDATION ---------------- //

export function validateDoctorLogin(formData) {
  const errors = [];
  const { email, password } = formData || {};

  if (!email || !email.trim()) {
    errors.push("Email is required.");
  } else if (!EMAIL_REGEX.test(email.trim())) {
    errors.push("Please enter a valid email address.");
  }

  if (!password || !password.trim()) {
    errors.push("Password is required.");
  } else if (password.length < 4) {
    errors.push("Password should be at least 4 characters.");
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
}

// ---------------- AI CHAT MESSAGE VALIDATION ---------------- //

export function validateAIMessage(message) {
  const errors = [];
  const text = (message || "").trim();

  if (!text) {
    errors.push("Please type your symptoms or question.");
  } else if (text.length < 5) {
    errors.push("Message is too short. Add a bit more detail.");
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
}

// ---------------- IMAGE FILE VALIDATION ---------------- //

export function validateImageFile(file) {
  const errors = [];

  if (!file) {
    errors.push("Please select an image file.");
  } else {
    // Type check
    if (!file.type.startsWith("image/")) {
      errors.push("File must be an image (jpg, png, etc.).");
    }

    // Size check
    const maxBytes = MAX_IMAGE_SIZE_MB * 1024 * 1024;
    if (file.size > maxBytes) {
      errors.push(`Image must be smaller than ${MAX_IMAGE_SIZE_MB} MB.`);
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
}

File: backend/app/utils/validators.py
# validators.py placeholder
# app/utils/validators.py

"""
Centralized validation utilities for the telemedicine app.

Used by:
- auth_routes.py (student & doctor login)
- places_routes.py (optional validation)
- future forms (e.g., symptom submissions)

All validators should return:
    (bool, message)
or raise a ValueError when used in service-level code.
"""

from typing import Dict, List, Tuple, Optional

# Allowed constants (to keep consistent with other files)
ALLOWED_HOSTELS = {
    "Boys Hostel 1",
    "Boys Hostel 2",
    "Girls Hostel 1",
    "Girls Hostel 2",
}

ALLOWED_LEVELS = {"UG", "PG"}

ALLOWED_PLACE_TYPES = {"hospital", "medical_shop"}


# ---------------------- Core Student Validation ---------------------- #

def validate_student_login_payload(data: Dict) -> Tuple[bool, List[str]]:
    """
    Validate a student quick login payload.

    Returns:
        (True, []) if valid,
        (False, [error messages]) if invalid.
    """

    errors = []

    name = (data.get("name") or "").strip()
    level = (data.get("level") or "").strip().upper()
    hostel = (data.get("hostel") or "").strip()
    branch = (data.get("branch") or "").strip()
    year = data.get("year")
    department = (data.get("department") or "").strip()

    # Name
    if not name:
        errors.append("Name is required.")

    # Level (UG / PG)
    if level not in ALLOWED_LEVELS:
        errors.append("Level must be 'UG' or 'PG'.")

    # Hostel
    if hostel not in ALLOWED_HOSTELS:
        errors.append("Hostel is invalid or missing.")

    # UG validation
    if level == "UG":
        if not branch:
            errors.append("Branch is required for UG students.")
        try:
            year_int = int(year)
            if year_int < 1 or year_int > 4:
                errors.append("Year must be between 1 and 4 for UG students.")
        except (ValueError, TypeError):
            errors.append("Year must be a number for UG students.")

    # PG validation
    if level == "PG":
        if not department:
            errors.append("Department is required for PG students.")

    return (len(errors) == 0, errors)


# ---------------------- Doctor Login Validation ---------------------- #

def validate_doctor_login_payload(data: Dict) -> Tuple[bool, List[str]]:
    """
    Validate a doctor login payload.

    Required:
        - email
        - password
    """
    errors = []
    email = (data.get("email") or "").strip().lower()
    password = (data.get("password") or "").strip()

    if not email:
        errors.append("Email is required.")
    elif "@" not in email:
        errors.append("Invalid email format.")

    if not password:
        errors.append("Password is required.")

    return (len(errors) == 0, errors)


# ---------------------- Places Validation ---------------------- #

def validate_place_payload(data: Dict) -> Tuple[bool, List[str]]:
    """
    Validate the JSON body for creating a new place (hospital or medical shop).

    Expected:
    {
      "name": "Apollo Pharmacy",
      "type": "medical_shop",
      "address": "Main Road",
      "hostel_tag": "Boys Hostel 1",
      ...
    }
    """
    errors = []

    name = (data.get("name") or "").strip()
    place_type = (data.get("type") or "").strip().lower()
    address = (data.get("address") or "").strip()
    hostel_tag = (data.get("hostel_tag") or "").strip()

    if not name:
        errors.append("Place name is required.")
    if place_type not in ALLOWED_PLACE_TYPES:
        errors.append("Type must be either 'hospital' or 'medical_shop'.")
    if not address:
        errors.append("Address is required.")
    if hostel_tag not in ALLOWED_HOSTELS:
        errors.append("hostel_tag must be one of the known hostels.")

    return (len(errors) == 0, errors)


# ---------------------- Generic Helpers ---------------------- #

def validate_nonempty_fields(data: Dict, required_fields: List[str]) -> Tuple[bool, List[str]]:
    """
    Simple helper to check if required fields are non-empty in a dict.
    """
    errors = []
    for field in required_fields:
        if not (data.get(field) or "").strip():
            errors.append(f"{field} is required.")
    return (len(errors) == 0, errors)


def is_valid_hostel(hostel: str) -> bool:
    """
    Return True if hostel name is allowed.
    """
    return hostel in ALLOWED_HOSTELS


def is_valid_level(level: str) -> bool:
    """
    Return True if level is UG or PG.
    """
    return level.upper() in ALLOWED_LEVELS
